Sources/BraintreeCoreSwift/BTHTTP.swift

```diff
@@ -0,0 +1,557 @@
+    }
+
+    @objc(DELETE:parameters:completion:)
+    public func delete(_ path: String, parameters: NSDictionary? = nil, completion: @escaping RequestCompletion) {
+        httpRequest(method: "DELETE", path: path, parameters: parameters, completion: completion)
+    }
+
+    // MARK: - HTTP Method Helpers
+
+    func httpRequestWithCaching(
+        method: String?,
+        path: String?,
+        parameters: NSDictionary? = [:],
+        completion: RequestCompletion?
+    ) {
+        createRequest(method: method, path: path, parameters: parameters) { request, error in
+            guard let request = request else {
+                self.handleRequestCompletion(data: nil, request: nil, shouldCache: false, response: nil, error: error, completion: completion)
+                return
+            }
+
+            var cachedResponse: CachedURLResponse? = URLCache.shared.cachedResponse(for: request) ?? nil
+
+            if self.cacheDateValidator.isCacheInvalid(cachedResponse ?? nil) {
+                URLCache.shared.removeAllCachedResponses()
+                cachedResponse = nil
+            }
+
+            // The increase in speed of API calls with cached configuration caused an increase in "network connection lost" errors.
+            // Adding this delay allows us to throttle the network requests slightly to reduce load on the servers and decrease connection lost errors.
+            DispatchQueue.main.asyncAfter(deadline: .now() + 0.1) {
+                if cachedResponse != nil {
+                    self.handleRequestCompletion(data: cachedResponse?.data, request: nil, shouldCache: false, response: cachedResponse?.response, error: nil, completion: completion)
+                } else {
+                    let task: URLSessionTask? = self.session?.dataTask(with: request) { [weak self] data, response, error in
+                        self?.handleRequestCompletion(data: data, request: request, shouldCache: true, response: response, error: error, completion: completion)
+                    }
+
+                    task?.resume()
+                }
+            }
+        }
+    }
+
+    func httpRequest(
+        method: String?,
+        path: String?,
+        parameters: NSDictionary? = [:],
+        completion: RequestCompletion?
+    ) {
+        createRequest(method: method, path: path, parameters: parameters) { request, error in
+            guard let request = request else {
+                self.handleRequestCompletion(data: nil, request: nil, shouldCache: false, response: nil, error: error, completion: completion)
+                return
+            }
+
+            let task: URLSessionTask? = self.session?.dataTask(with: request) { [weak self] data, response, error in
+                self?.handleRequestCompletion(data: data, request: request, shouldCache: false, response: response, error: error, completion: completion)
+            }
+
+            task?.resume()
```

Comment:
I think it would be safer to not allow an optional `URLSessionTask` and to call the completion with an error if it is.